src/QueryBuilder.php

```diff
@@ -467,9 +467,11 @@ public function setParameter(string|int $key, mixed $value, ParameterType|ArrayP
      *        )));
      * </code>
      *
-     * @psalm-param ArrayCollection<int, Parameter> $parameters
+     * @psalm-param ArrayCollection<TKey, Parameter> $parameters
      *
      * @return $this
+     *
+     * @template TKey of int
```

Comment:
What's the point? We're not using `TKey` anywhere else. Maybe we should instead tell `Psalm` that it's OK to pass a more specific collection, because `setParameters` is not going to mutate it. I believe that's what [`@psalm-external-mutation-free`](https://psalm.dev/articles/immutability-and-beyond) is about. Can you try annotating `setParameters` with that annotation instead?

Reply:
The point here is that `ArrayCollection<int, ...>` is too restrictive over the given `ArrayCollection` when inference comes from call-time parameters.

If you declare `new ArrayCollection(['a', 'b'])`, you get an `ArrayCollection<1|2, string>`, rather than an `ArrayCollection<int, string>`, which is then rejected by this `QueryBuilder` method.

This generalizes `TKey` over `int`, while before `int` specifically was required. It effectively says "we don't care what kind of `int` the key is, as long as it is within the bounds of this generic type.

Reply:
I wasn't wondering what the point of this change was, but whether there was the point of holding the type in a dedicated `TKey` name (since we don't use it). I would prefer a syntax allowing any subtype of `int` without having to come up with a name, but if that does not exist, let's go with that. I tried my own suggestion BTW, and it does not address the issue, I'm not sure why… I don't understand why exactly `ArrayCollection<1|2, string>` cannot be accepted where `ArrayCollection<int, string>` is. Well, I think thought that was because the method could alter the type, but that does not seem to be it.

Reply:
It's not a problem of altering the type. This problem occours because when you declare an `ArrayCollection<int, string>` you are forcing the templating inside `ArrayCollection` class to be restricted only to `int` type.
When you instanciate a new ArrayCollection passing an array <0,1> as parameters, the templating now is forced to only `int<0,1>` as type, which is more specific than the one you are asking for.
Using templating inside `setParameters()` allows the type to be inferred from the input parameter, with the only constrain of "extending" the int type.